hazelcast/src/hazelcast/client/spi.cpp

```diff
@@ -438,10 +444,10 @@ namespace hazelcast {
                         break;
                 }
 
-                for (std::unordered_set<LifecycleListener *>::iterator it = listeners.begin(); it != listeners.end(); ++it) {
-                    (*it)->stateChanged(lifecycleEvent);
+                for (auto &item: listeners) {
+                    auto& listener = item.second;
+                    listener.stateChanged(lifecycleEvent);
```

Comment:
should we consider making the interface a different function pointer line on_connected, on_shutdown, etc. ?